### PR TITLE
etcdserver: add kv Hash method (for testing purpose)

### PR DESCRIPTION
**Pull request by gyuho**
For https://github.com/coreos/etcd/issues/4380.

### CODE REVIEW

**Comment by gyuho** [comment]
/cc @xiang90 @heyitsanthony Please review. Will change https://github.com/coreos/etcd/issues/4380 after this PR.

Thanks,

File: etcdserver/etcdserverpb/rpc.proto [hunk 1]
```diff
@@ -229,6 +234,14 @@ message CompactionResponse {
   ResponseHeader header = 1;
 }
 
+message HashRequest {
+}
+
+message HashResponse {
+  bool succeeded = 1;
```
Comment (xiang90) [inline etcdserver/etcdserverpb/rpc.proto]: This field is useless when we already return error

File: etcdserver/api/v3rpc/key.go [hunk 2]
```diff
@@ -134,6 +134,14 @@ func (s *kvServer) Compact(ctx context.Context, r *pb.CompactionRequest) (*pb.Co
 	return resp, nil
 }
 
+func (s *kvServer) Hash(ctx context.Context, r *pb.HashRequest) (*pb.HashResponse, error) {
+	resp, err := s.kv.Hash(ctx, r)
+	if err != nil {
+		return nil, togRPCError(err)
```
Comment (xiang90) [inline etcdserver/api/v3rpc/key.go]: We need to handle this in to grpc error func

File: etcdserver/etcdserverpb/rpc.proto (hunk 1 above)
Comment (gyuho) [inline etcdserver/etcdserverpb/rpc.proto]: Ok will delete. Thanks.

File: etcdserver/api/v3rpc/key.go (hunk 2 above)
Comment (gyuho) [inline etcdserver/api/v3rpc/key.go]: @xiang90 Do you mean something else than `togRPCError`?

Comment (xiang90) [inline etcdserver/api/v3rpc/key.go]: `togRPCError` only handles known types, check 
https://github.com/coreos/etcd/blob/master/etcdserver/api/v3rpc/key.go#L279-L291. Other types will return as internal error type. I am not sure what you want to achieve here. Return as internal error type is probably just fine.

File: etcdserver/etcdserverpb/rpc.proto [hunk 3]
```diff
@@ -30,6 +30,11 @@ service KV {
   // Compact compacts the event history in etcd. User should compact the
   // event history periodically, or it will grow infinitely.
   rpc Compact(CompactionRequest) returns (CompactionResponse) {}
+
+  // Hash returns the hash of local KV state for consistency checking purpose.
+  // This is for internal testing. Do not use this in production when there
```
Comment (xiang90) [inline etcdserver/etcdserverpb/rpc.proto]: This is designed for testing purpose (not necessarily internal)

File: etcdserver/etcdserverpb/rpc.proto [hunk 4]
```diff
@@ -82,16 +87,16 @@ message ResponseHeader {
 
 message RangeRequest {
   enum SortOrder {
-	NONE = 0; // default, no sorting
-	ASCEND = 1; // lowest target value first
-	DESCEND = 2; // highest target value first
+  NONE = 0; // default, no sorting
```
Comment (xiang90) [inline etcdserver/etcdserverpb/rpc.proto]: why this is changed? can we change back?

File: etcdserver/v3demo_server.go [hunk 5]
```diff
@@ -88,6 +89,13 @@ func (s *EtcdServer) Compact(ctx context.Context, r *pb.CompactionRequest) (*pb.
 	return result.resp.(*pb.CompactionResponse), result.err
 }
 
+func (s *EtcdServer) Hash(ctx context.Context, r *pb.HashRequest) (*pb.HashResponse, error) {
+	h, err := s.be.Hash()
+	return &pb.HashResponse{
```
Comment (xiang90) [inline etcdserver/v3demo_server.go]: return (nil, err) when there is an error.

**Comment by xiang90** [comment]
LGTM

File: etcdserver/api/v3rpc/key.go (hunk 2 above)
Comment (gyuho) [inline etcdserver/api/v3rpc/key.go]: Ok will change it to return the internal error: `grpc.Errorf(codes.Internal, err.Error())`

Comment (xiang90) [inline etcdserver/api/v3rpc/key.go]: well... i mean `togRPCError` already convert it to internal error if this is an unknown type to the function. Please check the code of `togRPCError`

**Comment by gyuho** [comment]
@xiang90 Just fixed all. Will merge after test passes. Thanks.

**Comment by gyuho** [comment]
Test failure is not related to this change. `go test -run TestV3Hash` passes.

**Pull request merged**